### PR TITLE
CUDA: Add function to get SASS for kernels

### PR DESCRIPTION
**Pull request by gmarkall**
This enables calling a function `inspect_sass()` on kernels, similarly to how `inspect_asm()` provides the PTX code. It is not yet implemented for device functions, because device functions don't appear in the SASS when compiled alone without Relocatable Device Code (RDC) on, which is not presently supported by Numba.

The SASS disassembly is created by calling out to `nvdisasm`, which is required to be on the PATH.

### CODE REVIEW

**Comment by esc** [comment]
BFID: `numba_smoketest_cuda_79`

**Comment by esc** [comment]
The windows and linux builds for Python 3.6 failed on this one.

Windows:

```
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - ======================================================================
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - ERROR: test_inspect_sass_eager (numba.cuda.tests.cudapy.test_inspect.TestInspect)
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - ----------------------------------------------------------------------
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - Traceback (most recent call last):
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\site-packages\numba\cuda\tests\cudapy\test_inspect.py", line 98, in test_inspect_sass_eager
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     self._test_inspect_sass(add, 'add', add.inspect_sass())
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\site-packages\numba\cuda\compiler.py", line 683, in inspect_sass
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     return self._func.get_sass()
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\site-packages\numba\cuda\compiler.py", line 541, in get_sass
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     return disassemble_cubin(self.cubins[device.id])
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\site-packages\numba\cuda\compiler.py", line 147, in disassemble_cubin
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     capture_output=True)
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\subprocess.py", line 423, in run
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     with Popen(*popenargs, **kwargs) as process:
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - TypeError: __init__() got an unexpected keyword argument 'capture_output'
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - 
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - ======================================================================
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - ERROR: test_inspect_sass_lazy (numba.cuda.tests.cudapy.test_inspect.TestInspect)
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - ----------------------------------------------------------------------
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - Traceback (most recent call last):
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\site-packages\numba\cuda\tests\cudapy\test_inspect.py", line 113, in test_inspect_sass_lazy
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     self._test_inspect_sass(add, 'add', add.inspect_sass(signature))
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\site-packages\numba\cuda\compiler.py", line 954, in inspect_sass
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     return self.definitions[(cc, signature)].inspect_sass()
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\site-packages\numba\cuda\compiler.py", line 683, in inspect_sass
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     return self._func.get_sass()
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\site-packages\numba\cuda\compiler.py", line 541, in get_sass
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     return disassemble_cubin(self.cubins[device.id])
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\site-packages\numba\cuda\compiler.py", line 147, in disassemble_cubin
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     capture_output=True)
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -   File "F:\ci_envs\64\Miniconda3\envs\testenv_a2d43c9c-5007-4446-be28-5f2295392dd7\lib\subprocess.py", line 423, in run
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO -     with Popen(*popenargs, **kwargs) as process:
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - TypeError: __init__() got an unexpected keyword argument 'capture_output'
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - 
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - ----------------------------------------------------------------------
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - Ran 642 tests in 150.067s
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - 
[2020-07-29 04:23:52,879] {bash_operator.py:137} INFO - FAILED (errors=2, skipped=17)
```

Linux:

```
[2020-07-29 08:51:29,875] {docker_operator.py:265} INFO - ======================================================================
[2020-07-29 08:51:29,875] {docker_operator.py:265} INFO - ERROR: test_inspect_sass_nvdisasm_missing (numba.cuda.tests.cudapy.test_inspect.TestInspect)
[2020-07-29 08:51:29,876] {docker_operator.py:265} INFO - ----------------------------------------------------------------------
[2020-07-29 08:51:29,876] {docker_operator.py:265} INFO - Traceback (most recent call last):
[2020-07-29 08:51:29,876] {docker_operator.py:265} INFO - File "/opt/conda/envs/testenv_53b2f702-6bb8-473f-937f-94f305f2b223/lib/python3.6/site-packages/numba/cuda/tests/cudapy/test_inspect.py", line 123, in test_inspect_sass_nvdisasm_missing
[2020-07-29 08:51:29,876] {docker_operator.py:265} INFO - f.inspect_sass()
[2020-07-29 08:51:29,876] {docker_operator.py:265} INFO - File "/opt/conda/envs/testenv_53b2f702-6bb8-473f-937f-94f305f2b223/lib/python3.6/site-packages/numba/cuda/compiler.py", line 683, in inspect_sass
[2020-07-29 08:51:29,876] {docker_operator.py:265} INFO - return self._func.get_sass()
[2020-07-29 08:51:29,876] {docker_operator.py:265} INFO - File "/opt/conda/envs/testenv_53b2f702-6bb8-473f-937f-94f305f2b223/lib/python3.6/site-packages/numba/cuda/compiler.py", line 541, in get_sass
[2020-07-29 08:51:29,876] {docker_operator.py:265} INFO - return disassemble_cubin(self.cubins[device.id])
[2020-07-29 08:51:29,876] {docker_operator.py:265} INFO - File "/opt/conda/envs/testenv_53b2f702-6bb8-473f-937f-94f305f2b223/lib/python3.6/site-packages/numba/cuda/compiler.py", line 147, in disassemble_cubin
[2020-07-29 08:51:29,877] {docker_operator.py:265} INFO - capture_output=True)
[2020-07-29 08:51:29,877] {docker_operator.py:265} INFO - File "/opt/conda/envs/testenv_53b2f702-6bb8-473f-937f-94f305f2b223/lib/python3.6/subprocess.py", line 423, in run
[2020-07-29 08:51:29,877] {docker_operator.py:265} INFO - with Popen(*popenargs, **kwargs) as process:
[2020-07-29 08:51:29,877] {docker_operator.py:265} INFO - TypeError: __init__() got an unexpected keyword argument 'capture_output'
[2020-07-29 08:51:29,877] {docker_operator.py:265} INFO - 
[2020-07-29 08:51:29,877] {docker_operator.py:265} INFO - ----------------------------------------------------------------------
[2020-07-29 08:51:29,877] {docker_operator.py:265} INFO - Ran 642 tests in 135.212s
[2020-07-29 08:51:29,878] {docker_operator.py:265} INFO - 
[2020-07-29 08:51:29,878] {docker_operator.py:265} INFO - FAILED (errors=1, skipped=9)
```

**Comment by gmarkall** [comment]
@esc Could you smoketest this one again please?

**Comment by esc** [comment]
BFID: `numba_smoketest_cuda_82`

**Comment by gmarkall** [comment]
@esc many thanks! @sklam would you be happy to review this?

**Comment by gmarkall** [comment]
@esc conflicts fixed - given that a bit of a refactoring of the CUDA kernel and dispatcher mechanisms just got merged, could this have another buildfarm run please?

**Comment by esc** [comment]
BFID: `numba_smoketest_cuda_84`

File: numba/cuda/compiler.py [hunk 1]
```diff
@@ -547,6 +590,14 @@ def inspect_asm(self):
         '''
         return self._func.ptx.get().decode('ascii')
 
+    def inspect_sass(self):
```
Comment (sklam) [inline numba/cuda/compiler.py:593]: Should update the docs at  https://github.com/numba/numba/blame/master/docs/source/cuda-reference/kernel.rst#L60

File: numba/cuda/compiler.py [hunk 2]
```diff
@@ -130,6 +132,35 @@ def compile_ptx_for_current_device(pyfunc, args, debug=False, device=False,
                        fastmath=fastmath, cc=cc, opt=True)
 
 
+def disassemble_cubin(cubin):
+    # nvdisasm only accepts input from a file, so we need to write out to a
+    # temp file and clean up afterwards.
+    fd = None
+    fname = None
+    try:
+        fd, fname = tempfile.mkstemp()
```
Comment (sklam) [inline numba/cuda/compiler.py:141]: Using `NamedTemporaryFile` here would simplify the code and make it less error prone. 
For reference: https://github.com/numba/numba/blob/master/numba/misc/inspection.py#L20-L22

**Comment by esc** [comment]
BF passed again

File: numba/cuda/compiler.py (hunk 2 above)
Comment (gmarkall) [reply numba/cuda/compiler.py:141]: I can't see how to use `NamedTemporaryFile` here in a way that would simplify the logic or make it less error-prone: [its documentation](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) states that on Windows, the file can't be opened a second time whilst it is still open (as would be needed to open it with `nvdisasm`). So I'd have to close the file, open it with nvdisasm, and then finally delete it anyway.

- Am I missing something? How should the simpler version with `NamedTemporaryFile` work?
- If I'm not missing something, does the code you referenced suffer from the same problem? I had a look for tests that would hit it, but I couldn't find any.

File: numba/cuda/compiler.py (hunk 1 above)
Comment (gmarkall) [reply numba/cuda/compiler.py:593]: I've now added `inspect_sass` to the docs - thanks for catching this!

File: numba/cuda/compiler.py (hunk 2 above)
Comment (sklam) [reply numba/cuda/compiler.py:141]: Good catch on the can't-open-file-twice limitation. I wasn't aware of that. And for the `disassemble_elf_to_cfg`, I don't think anyone has tested it on windows.

**Pull request merged**